app/src/input.rs

```diff
@@ -1,32 +1,285 @@
+    for (x, y, z) in gforce_vec {
+        x_sum += x;
+        y_sum += y;
+        z_sum += z;
+    }
+    match gforce_vec.len().to_f32() {
+        Some(len) => (x_sum / len, y_sum / len, z_sum / len),
+        None => (0.0, 0.0, 0.0),
+    }
+}
+
+pub struct Motion {
+    pub direction: Direction,
+    pub started: bool,
+    pub ended: bool,
+    _minimal_steps: usize,
+}
+
+impl Motion {
+    pub fn new(direction: Direction) -> Self {
+        Motion {
+            direction,
+            started: true, // true only at the first iteration of the motion
+            ended: false,  // true only at the last iteration of the motion
+            _minimal_steps: 7,
+        }
+    }
+
+    pub fn create_if_needed(
+        neutral_gforce: (f32, f32, f32),
+        movement_gforce: (f32, f32, f32),
+        plots_holder: &mut PlotsHolder,
+    ) -> Option<Motion> {
+        let (total_gforce, corrected_gforce) = process_gforce(neutral_gforce, movement_gforce);
+        return if total_gforce >= 1.75 {
+            let dir = find_direction(corrected_gforce);
+            println!(
+                "Motion started: {:?} {:?} {}",
+                dir, corrected_gforce, total_gforce
+            );
+            if plots_holder.motion_plot {
+                plots_holder.add_measurement(
+                    "movement",
+                    vec!["x", "y", "z", "total"],
+                    vec![
+                        corrected_gforce.0,
+                        corrected_gforce.1,
+                        corrected_gforce.2,
+                        total_gforce,
+                    ],
+                );
+            }
+            let m = Motion::new(dir);
+            Some(m)
+        } else {
+            None
+        };
+    }
+
+    pub fn update(
+        &mut self,
+        neutral_gforce: (f32, f32, f32),
+        movement_gforce: (f32, f32, f32),
+        plots_holder: &mut PlotsHolder,
+    ) {
+        self.started = false;
+        let (total_gforce, corrected_gforce) = process_gforce(neutral_gforce, movement_gforce);
+        plots_holder.add_measurement(
+            "movement",
+            vec!["x", "y", "z", "total"],
+            vec![
+                corrected_gforce.0,
+                corrected_gforce.1,
+                corrected_gforce.2,
+                total_gforce,
+            ],
+        );
+        if self._minimal_steps > 0 {
+            self._minimal_steps -= 1;
+            return;
+        }
+        if total_gforce < 1.0 {
+            self.ended = true;
+            println!("Motion ended: {:?} {}", self.direction, total_gforce);
+            if plots_holder.motion_plot {
+                plots_holder.plots_to_logs()
+            }
+        }
+    }
+
+    pub fn to_motion_control(&self) -> MotionControl {
+        MotionControl {
+            direction: self.direction,
+            started: self.started,
+            ended: self.ended,
+        }
+    }
+}
+
+fn process_gforce(
```

Comment:
All this code is black magic to me, but I'm also glazing over it a bit. If you think it's possible please add method comments that briefly explain their purpose. Might help you too if you're sleep deprived (hopefully not, but you know) and working towards the deadline ;)